book/src/rfcs/RFC0009-Cycle-recovery.md

```diff
@@ -0,0 +1,126 @@
+# Description/title
+
+## Metadata
+
+* Author: nikomatsakis
+* Date: 2021-10-31
+* Introduced in: https://github.com/salsa-rs/salsa/pull/285
+
+## Summary
+
+* Expand `Cancelled` type (introduced in [RFC 7]) to carry information about *why* cancellation occurred.
+* Introduce `Cycle` type that carries information about a cycle.
+* Modify cycle recovery to take a `&Cycle`.
+* Report cycle participants in a deterministic order.
+
+[RFC 7]: ./RFC0007-Opinionated-Cancelation.md
+
+## Motivation
+
+Cycle recovery has been found to have some subtle bugs that could lead to panics. Furthermore, the existing cycle recovery APIs give limited and non-deterministic information. This RFC tweaks the user exposed APIs to correct these shortcomings. It also describes a major overhaul of how cycles are handled internally.
+
+## User's guide
+
+By default, when Salsa detects a cycle in the computation graph, Salsa will panic with a `Cancelled::UnexpectedCycle` as the panic value. This variant includes a `Cycle` structure that describes the cycle, which can be useful for diagnosing what went wrong. You can catch panics like these and inspect the cycle like so:
+
+```rust
+match Cancelled::catch(|| {
+    db.do_query()
+}) {
+    Ok(v) => v,
+    Err(Cancelled::UnexpectedCycle(cycle)) => {
+        eprintln("cycle results: {:#?}", cycle.debug(db))
+    }
+    Err(_) => {
+        /* ... */        
+    }
+}
```

Comment:
Need to read the proposal in more detail later, but this approach of treating unexpected cycle as a variant of cancelled seems "obviously wrong" to me, so I wonder if I might be missing something :) 

The old `Cancelled` panic wasn't an error -- it was serendipitous success. The new `Cancelled::UnexpectedCycle` seems like it signals an error, specifically, a programmers error, a bug in a program. That is, it seems that we are now mixing two very distinct cases in the same `Cancelled::catch` API:

* recoverable (even successful) "normal operation"
* erroneous "bug in the program", which is unrecoverable 

I would expect that, for unexpected cycles, we'd just `panic!` with a message printing out the cycle, or, if we care a lot about giving access to `Cycle` for debugging purposes, with throwing the `Cycle` itself.

Reply:
Yeah, I might have chosen wrong here. I debated about precisely this question and I was thinking that, in the future, I expect to move salsa so that it manages a pool of helper threads, and that (for those threads) irrecoverable cycles would want to be caught and forwarded, so it was kind of "akin" to cancellation, but really I agree with you that they are distinct things. I can change the code to just throw the `Cycle`, as suggested.

That said, this is not the "interesting" part of the change -- we used to panic with a string, we now panic with a useful value. The more interesting part is how we treat *recoverable* cycles. There we unwind with `CycleParticipant` and catch it internally.

Reply:
The latest commit throws `Cycle` values directly.